### PR TITLE
Epoch processing spec tests

### PR DESCRIPTION
**Pull request by terencechain**
Current status: 

|                                   |                         |
|----------------------------|----------------------|
| crosslinks | PASS |
| registry_updates | PASS |

https://github.com/ethereum/eth2.0-specs/tree/dev/specs/test_formats/epoch_processing

### CODE REVIEW

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 1]
```diff
@@ -20,33 +20,18 @@ var totalActiveBalanceCache = cache.NewActiveBalanceCache()
 //    Return the combined effective balance of the ``indices``. (1 Gwei minimum to avoid divisions by zero.)
 //    """
 //    return max(sum([state.validator_registry[index].effective_balance for index in indices]), 1)
-func TotalBalance(state *pb.BeaconState, indices []uint64) (uint64, error) {
-	epoch := CurrentEpoch(state)
-	total, err := totalBalanceCache.TotalBalanceInEpoch(epoch)
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]: Removing total balance cache usage for `TotalBalance`. It was not implemented correctly and it's not needed. The input should have been `indices` not `epoch`

File: beacon-chain/core/epoch/epoch_processing.go [hunk 2]
```diff
@@ -631,14 +627,20 @@ func winningCrosslink(state *pb.BeaconState, shard uint64, epoch uint64) (*pb.Cr
...
 		if err != nil {
-			return nil, nil, fmt.Errorf("could not hash crosslink from state: %v", err)
+			return nil, nil, fmt.Errorf("could not hash tree root crosslink from state: %v", err)
 		}
-		if proto.Equal(cFromState, a.Data.Crosslink) || bytes.Equal(h[:], a.Data.Crosslink.ParentRoot) {
+		cFromAttRoot, err := ssz.HashTreeRoot(a.Data.Crosslink)
+		if err != nil {
+			return nil, nil, fmt.Errorf("could not hash tree root crosslink from attestation: %v", err)
+		}
+		if bytes.Equal(cFromStateRoot[:], cFromAttRoot[:]) || bytes.Equal(cFromStateRoot[:], a.Data.Crosslink.ParentRoot) {
```
Comment (rauljordan) [inline beacon-chain/core/epoch/epoch_processing.go]: This is very verbose, can we simplify into clear variables?

File: beacon-chain/core/epoch/epoch_processing.go [hunk 3]
```diff
@@ -631,14 +627,20 @@ func winningCrosslink(state *pb.BeaconState, shard uint64, epoch uint64) (*pb.Cr
 	// Filter out shard crosslinks with correct current or previous crosslink data.
 	for _, a := range shardAtts {
 		cFromState := state.CurrentCrosslinks[shard]
-		h, err := hashutil.HashProto(cFromState)
+
+		cFromStateRoot, err := ssz.HashTreeRoot(cFromState)
```
Comment (rauljordan) [inline beacon-chain/core/epoch/epoch_processing.go]: cFromStateRoot isn't a very meaningful name, can we improve it and make it clear that the c refers to a crosslink? A better name would help

File: beacon-chain/core/epoch/spectest/epoch_processing_crosslink_test.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/ghodss/yaml"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/epoch"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/params/spectest"
+	"github.com/prysmaticlabs/prysm/shared/testutil"
+)
+
+func runCrosslinkProcessingTests(t *testing.T, filename string) {
+	file, err := ioutil.ReadFile(filename)
```
Comment (rauljordan) [inline beacon-chain/core/epoch/spectest/epoch_processing_crosslink_test.go]: Can the logic from line 17-25 be abstracted in some common method?

File: beacon-chain/core/epoch/spectest/epoch_processing_registry_test.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
+package spectest
+
+import (
+	"github.com/prysmaticlabs/prysm/shared/testutil"
```
Comment (rauljordan) [inline beacon-chain/core/epoch/spectest/epoch_processing_registry_test.go]: Run goimports

File: beacon-chain/core/helpers/rewards_penalties_test.go [hunk 6]
```diff
@@ -12,10 +12,8 @@ func TestTotalBalance_OK(t *testing.T) {
 		{EffectiveBalance: 32 * 1e9}, {EffectiveBalance: 40 * 1e9},
 	}}
 
-	balance, err := TotalBalance(state, []uint64{0, 1, 2, 3})
-	if err != nil {
-		t.Fatal(err)
-	}
+	balance := TotalBalance(state, []uint64{0, 1, 2, 3})
+
 	if balance != 127*1e9 {
```
Comment (rauljordan) [inline beacon-chain/core/helpers/rewards_penalties_test.go]: Magic number? This is the sort of stuff that breaks future tests and makes things hard to debug

File: beacon-chain/core/helpers/rewards_penalties_test.go [hunk 7]
```diff
@@ -12,10 +12,8 @@ func TestTotalBalance_OK(t *testing.T) {
...
 		t.Errorf("Incorrect TotalEffectiveBalance. Wanted: 127, got: %d", balance)
```
Comment (rauljordan) [inline beacon-chain/core/helpers/rewards_penalties_test.go]: Same here

File: beacon-chain/core/epoch/spectest/epoch_processing_crosslink_test.go (hunk 4 above)
Comment (terencechain) [reply beacon-chain/core/epoch/spectest/epoch_processing_crosslink_test.go]: will address this in a separate PR, better to do it all at once with the outstanding spec testing PRs.

**Comment by terencechain** [comment]
reviews addressed @rauljordan

**Pull request merged**